readme.md
# tg_api_gen

library that generates realistic device_model, app_verion, and system_version (android and ios) for telegram userbots

there is a very high risk to get ban from telegram, please don't use with your main account

## installation
```bash
pip install tg_api_gen pyrogram tgcrypto
```

## how to use

just see what data generated:

```python
import tg_api_gen

print(tg_api_gen.generate_random())
 ```

real code example - [usage.py](usage.py)

## changelog

see [changelog.md](changelog.md)

## credits

thanks to:
- [opentele](https://github.com/thedemons/opentele)
- [pyrogram](https://github.com/pyrogram/pyrogram)
- [telethon](https://github.com/LonamiWebs/Telethon)

## license

license is gnu agpl 3 - [gnu.org/licenses/agpl-3.0.en.html](https://gnu.org/licenses/agpl-3.0.en.html)



changelog.md
# 23.1.0

- added license (agpl3)
- added usage.py
- added changelog.md
- added iphone 15 model
- added ios 17
- added new minor ios 15 and 16 versions
- adden new telegram ios and telegram android versions
- edited readme.md

# 23.0.1

- fixed typos

# 23.0.0

- init release
- implemented android and ios generation



main.py
import rich.console
from random import choice
from dataclasses import dataclass

c = rich.console.Console()


def get_device_and_ios() -> list:
    return choice(
        [
            [
                ['6', '6 Plus', '6S', '6S Plus'],
                [12, 15],
            ],
            [
                ['7', '7 Plus'],
                [12, 15],
            ],
            [
                ['8', '8 Plus'],
                [12, 16],
            ],
            [
                ['X', 'XS', 'XS Max', 'XR'],
                [12, 16],
            ],
            [
                ['11', '11 Pro', '11 Pro Max'],
                [12, 16],
            ],
            [
                ['12', '12 mini', '12 Pro', '12 Pro Max'],
                [14, 16]
            ],
            [
                ['13', '13 mini', '13 Pro', '13 Pro Max'],
                [15, 16]
            ],
            [
                ['14', '14 Plus', '14 Pro', '14 Pro Max'],
                [16]
            ],
        ]
    )


def get_ios_major(
    major_list: list,
):
    if len(major_list) == 1:
        return major_list[0]
    else:
        return choice(
            range(
                major_list[0],
                major_list[1] + 1,
            )
        )


def get_ios(
    major: int
) -> str:
    versions: list[list[int]] = [
        [16, 0, 3],
        [16, 1, 2],
        [16, 2, 0],
        [16, 3, 1],
        [16, 4, 1],
        [16, 5, 1],
        [15, 0, 2],
        [15, 1, 1],
        [15, 2, 1],
        [15, 3, 1],
        [15, 4, 1],
        [15, 5, 6],
        [15, 6, 1],
        [15, 7, 7],
        [14, 0, 1],
        [14, 1, 0],
        [14, 2, 1],
        [14, 3, 0],
        [14, 4, 2],
        [14, 5, 1],
        [14, 6, 0],
        [14, 7, 1],
        [14, 8, 1],
        [13, 1, 3],
        [13, 2, 3],
        [13, 3, 1],
        [13, 4, 1],
        [13, 5, 1],
        [13, 6, 1],
        [13, 7, 0],
        [12, 0, 1],
        [12, 1, 4],
        [12, 2, 0],
        [12, 3, 2],
        [12, 4, 9],
        [12, 5, 7],
    ]
    return version_to_str(
        *choice(versions)
    )


def version_to_str(
    major: int,
    minor: int,
    patch: int,
) -> str:
    patch = choice(range(patch + 1))
    if minor and patch:
        return f'{major}.{minor}.{patch}'
    elif minor:
        return f'{major}.{minor}'
    else:
        return str(major)


def get_app_version() -> str:
    versions = [
        [9, 6, 3],
        [9, 5, 3],
        [9, 4, 2],
    ]
    return version_to_str(
        *choice(versions)
    )


def generage_ios() -> dict:
    devices, ioses = get_device_and_ios()
    ios_major = get_ios_major(ioses)

    app_version = get_app_version()
    device_model = 'iPhone ' + choice(devices)
    system_version = get_ios(ios_major)

    return {
        'api_id': 1,
        'api_hash': 'b6b154c3707471f5339bd661645ed3d6',
        'app_version': app_version,
        'device_model': device_model,
        'system_version': system_version,
    }



build.py
#!/bin/python

import shutil
import toml
import sys
import os
from pathlib import Path
from tg_api_gen import (
    proj_version,
    proj_name,
)


app_path = Path(
    __file__
).parent.resolve()
pypr_path = Path(
    f'{app_path}/pyproject.toml'
)
dist_path = Path(
    f'{app_path}/dist'
)
pypr_data = {
    'build-system': {
        'requires': ['hatchling'],
        'build-backend': 'hatchling.build',
    },
    'project': {
        'dependencies': [],
        'name': proj_name,
        'version': proj_version,
        'authors': [
            {
                'name': 'gmanka',
                'email': 'gmankab@gmail.com',
            },
        ],
        'description': 'lightweight skript that generates realistic device_model, app_verion, system_version for telegram',
        'readme': 'readme.md',
        'requires-python': '>=3.10',
        'classifiers': [
            'Programming Language :: Python :: 3',
            'Operating System :: OS Independent',
        ],
        'urls': {
            'Homepage': 'https://github.com/gmankab/tg_api_gen',
        }
    }
} 


with open(
    pypr_path,
    'w',
) as file:
    file.write(
        toml.dumps(
            pypr_data
        )
    )


shutil.rmtree(
    dist_path,
    ignore_errors = True,
)

os.system(f'{sys.executable} -m prettygit')



usage.py
import pyrogram.errors
import pyrogram.client
import tg_api_gen
import asyncio

async def authorise(client):
    phone = input('phone: ')
    code = await client.send_code(
        phone_number = phone
    )
    print(code)
    try:
        return await client.sign_in(
            phone_number = phone,
            phone_code_hash = code.phone_code_hash,
            phone_code = input('code: '),
        )
    except pyrogram.errors.SessionPasswordNeeded:
        return await client.check_password(
            password = input('password: ')
        )

async def main():
    client = pyrogram.client.Client(
        name = 'my_client_name',
        **tg_api_gen.generate_random(),
    )
    await client.connect()
    try:
        user = await client.get_me()
        print('user already authorised')
    except:
        print('user is not authorised yet, authorising...')
        user = await authorise(client)
    print(user)
    await client.send_message('me', 'hello world')
    await client.stop()

asyncio.run(main())



tg_api_gen/__init__.py
# license is gnu agpl 3 - gnu.org/licenses/agpl-3.0.en.html

import string
from pathlib import Path
from random import (
    choice,
    randint,
)


proj_name = 'tg_api_gen'
proj_version = '23.1.0'


def get_device_and_ios_major() -> tuple:
    devices_and_ioses = [
        [
            ['6', '6 Plus', '6S', '6S Plus'],
            [12, 15],
        ],
        [
            ['7', '7 Plus'],
            [12, 15],
        ],
        [
            ['8', '8 Plus'],
            [12, 16],
        ],
        [
            ['X', 'XS', 'XS Max', 'XR'],
            [12, 17],
        ],
        [
            ['11', '11 Pro', '11 Pro Max'],
            [12, 17],
        ],
        [
            ['12', '12 mini', '12 Pro', '12 Pro Max'],
            [14, 17],
        ],
        [
            ['13', '13 mini', '13 Pro', '13 Pro Max'],
            [15, 17],
        ],
        [
            ['14', '14 Plus', '14 Pro', '14 Pro Max'],
            [16, 17],
        ],
        [
            ['15', '15 Plus', '15 Pro', '15 Pro Max'],
            [17],
        ]
    ]
    devices, ios_majors = choice(devices_and_ioses)
    device_model: str = 'iPhone ' + choice(devices)
    if len(ios_majors) == 1:
        ios_major: int = ios_majors[0]
    else:
        ios_major = choice(
            range(
                ios_majors[0],
                ios_majors[1] + 1,
            )
        )
    return device_model, ios_major


def get_ios_major(
    major_list: list,
):
    if len(major_list) == 1:
        return major_list[0]
    else:
        return choice(
            range(
                major_list[0],
                major_list[1] + 1,
            )
        )


def get_device_model_android():
    android_devices_path = Path(
        __file__
    ).parent.resolve() / 'android_devices.txt'
    return choice(
        list(
            android_devices_path.open('r')
        )
    ).strip('\n')


def get_ios_system_version(
    major: int
) -> str:
    versions: dict = {
        17: [
            [17, 0, 3],
            [17, 1, 2],
            [17, 2, 1],
        ],
        16: [
            [16, 0, 3],
            [16, 1, 2],
            [16, 2, 0],
            [16, 3, 1],
            [16, 4, 1],
            [16, 5, 1],
            [16, 6, 7],
            [16, 7, 4],
        ],
        15: [
            [15, 0, 2],
            [15, 1, 1],
            [15, 2, 1],
            [15, 3, 1],
            [15, 4, 1],
            [15, 5, 6],
            [15, 6, 1],
            [15, 7, 7],
            [15, 8, 0]
        ],
        14: [
            [14, 0, 1],
            [14, 1, 0],
            [14, 2, 1],
            [14, 3, 0],
            [14, 4, 2],
            [14, 5, 1],
            [14, 6, 0],
            [14, 7, 1],
            [14, 8, 1],
        ],
        13: [
            [13, 1, 3],
            [13, 2, 3],
            [13, 3, 1],
            [13, 4, 1],
            [13, 5, 1],
            [13, 6, 1],
            [13, 7, 0],
        ],
        12: [
            [12, 0, 1],
            [12, 1, 4],
            [12, 2, 0],
            [12, 3, 2],
            [12, 4, 9],
            [12, 5, 7],
        ],
    }
    return version_to_str(
        *choice(
            versions[major]
        )
    )


def get_android_system_version():
    return choice(
        [
            '7 N MR1 (25)',
            '8 O MR1 (26)',
            '8.1 O MR1 (27)',
            '9 P (28)',
            '10 Q (29)',
            '11 (30)',
            '12 (31)',
            # '12L (32)',
            '13 (33)',
            'SDK 34',
        ]
    )


def version_to_str(
    major: int,
    minor: int,
    patch: int,
) -> str:
    patch = choice(range(patch + 1))
    if minor and patch:
        return f'{major}.{minor}.{patch}'
    elif minor:
        return f'{major}.{minor}'
    else:
        return str(major)


def get_ios_app_version() -> str:
    versions = [
        [9,  6, 3],
        [9,  5, 3],
        [9,  4, 2],
        [9,  5, 3],
        [9,  6, 6],
        [10, 0, 3],
        [10, 1, 2],
        [10, 2, 6],
        [10, 3, 1],
    ]
    return 'Telegram IOS ' + version_to_str(
        *choice(versions)
    )


def get_android_app_version() -> str:
    versions = [
        [9,  6, 7],
        [9,  5, 8],
        [9,  4, 9],
        [9,  5, 8],
        [9,  6, 7],
        [9,  7, 6],
        [10, 0, 9],
        [10, 1, 3],
        [10, 2, 9],
        [10, 3, 2],
    ]
    return 'Telegram Android ' + version_to_str(
        *choice(versions)
    )


def generate_ios() -> dict:
    app_version = get_ios_app_version()
    device_model, ios_major = get_device_and_ios_major()
    system_version = get_ios_system_version(ios_major)
    return {
        'api_id': 10840,
        'api_hash': '33c45224029d59cb3ad0c16134215aeb',
        'app_version': app_version,
        'device_model': device_model,
        'system_version': system_version,
    }


def generate_android() -> dict:
    app_version = get_android_app_version()
    device_model = get_device_model_android()
    system_version = get_android_system_version()
    return {
        'api_id': 6,
        'api_hash': 'eb06d4abfb49dc3eeb1aeb98ae0f581e',
        'app_version': app_version,
        'device_model': device_model,
        'system_version': system_version,
    }


def generate_random():
    return choice(
        [
            generate_android,
            generate_ios,
        ]
    )()


def generate_name() -> str:
    result = ''
    for _ in range(randint(5, 10)):
        result += choice(string.ascii_lowercase)
    return result

